## 1.)  Take the following survey:  [CS207 Introductory Survey](https://docs.google.com/forms/d/e/1FAIpQLSeqarqOVyLqMs-u8ZKCEmQ-ccPjtg3qqA-UlR_Gb3zf3pDenQ/viewform)

### The problems in this homework set are based upon an example taken from SICP in which we try to write a bank acount withdraw system:

In [1]:
def make_withdraw1(balance):
    """Return a withdraw function that draws down balance with each call."""
    def withdraw(amount):#you are ok if you use balance as a read-only property
        if amount > balance:
            return 'Insufficient funds'
        return balance - amount
    return withdraw

We'll define an initial balance, set up the initial account balance, and make a withdraw.

In [3]:
init_bal = 500
wd_from_init = make_withdraw1(init_bal)
wd_from_init(75)

425

Now our new balance is 425 currency units.  Let's make another withdraw.

In [4]:
wd_from_init(25)

475

This is totally wrong!  Our new balance went UP even though we took out more money.  Remember, writing it like we did, we have no access to balance directly. In other words, we want balance that was captured from the enclosing frame to be changeable, not unchangeable.

One simple way to fix this would be to reassign `balance` within the `widthdraw` function.  We'll try that out in the next problem.

## 2.). Create a new `make_widthdraw` function (maybe call it `make_withdraw2`) and reassign `balance` within the `withdraw` function.  What happens?

In [5]:
# Your code here

Now try again.

In [ ]:
wd_from_init = make_widthdraw2(init_bal)
wd_from_init(75)

If you did it right, you should have got an error. As the Python Execution Model says:

>If a name binding operation occurs anywhere within a code block, all uses of the name within the block are treated as references to the current block. This can lead to errors when a name is used within a block before it is bound. This rule is subtle. Python lacks declarations and allows name binding operations to occur anywhere within a code block. The local variables of a code block can be determined by scanning the entire text of the block for name binding operations.

What has happened is that balance is now being treated as a local variable in withdraw, since the assignment has created a "name" for it. And then it does not have a value.

We need to somehow tell Python that balance is *not* a local variable. The way to do this is the `nonlocal` keyword.

## 3.) Write a function `make_withdraw` using

## `nonlocal balance`

## as the first line of the inner `withdraw` function.

In [6]:
# Your code goes here.

In [ ]:
wd_from_init = make_withdraw(init_bal)
wd_from_init(75)
wd_from_init(25)